File: smart-farm-server/src/models/SensorData.js
const mongoose = require('mongoose');

const SensorDataSchema = new mongoose.Schema(
  {
    deviceId: { type: mongoose.Schema.Types.ObjectId, ref: 'Device', required: true, index: true },
    temperature: { type: Number },
    humidity: { type: Number },
    soilMoisture: { type: Number },
    lux: { type: Number },
    pH: { type: Number },
    timestamp: { type: Date, default: Date.now, index: true },
  },
  { minimize: true }
);

SensorDataSchema.index({ deviceId: 1, timestamp: -1 });

module.exports = mongoose.model('SensorData', SensorDataSchema);

File: code_frontend/lib/models/sensor_data.dart
class SensorData {
  final DateTime timestamp;
  final double? temperature;
  final double? humidity;
  final double? soilMoisture;
  final double? lux;

  SensorData({
    required this.timestamp,
    this.temperature,
    this.humidity,
    this.soilMoisture,
    this.lux,
  });

  factory SensorData.fromJson(Map<String, dynamic> j) => SensorData(
    timestamp: DateTime.parse(
      j['timestamp'] ?? j['createdAt'] ?? DateTime.now().toIso8601String(),
    ),
    temperature: j['temperature'] != null
        ? (j['temperature'] as num).toDouble()
        : null,
    humidity: j['humidity'] != null ? (j['humidity'] as num).toDouble() : null,
    soilMoisture: j['soilMoisture'] != null
        ? (j['soilMoisture'] as num).toDouble()
        : null,
    lux: j['lux'] != null ? (j['lux'] as num).toDouble() : null,
  );
}
